src/pages/RealmExplorer.js

```diff
@@ -3,201 +3,214 @@
+  const filteredData = objectData;
+  return { objectString, objectData, searchString, filterString, filteredData };
+};
 
-    if (data) {
-      this.setState({
-        data,
-        dataSource: dataSource.cloneWithRows(data),
-      });
+const getUpdatedState = (database, state) => {
+  const { objectString, objectData, searchString, filterString, filteredData } = state;
+  const newState = { searchString, filterString };
+  const updateObject = searchString !== objectString && OBJECT_TYPES.indexOf(searchString) >= 0;
+  newState.objectString = updateObject ? searchString : objectString;
+  newState.objectData = updateObject ? database.objects(newState.objectString) : objectData;
+  if (filterString === '') {
+    newState.filteredData = newState.objectData;
+  } else {
+    try {
+      newState.filteredData = newState.objectData.filtered(filterString);
+    } catch (err) {
+      newState.filteredData = updateObject ? newState.objectData : filteredData;
     }
   }
 
-  onSearchChange(searchTerm) {
-    const { database } = this.props;
-    const { dataSource } = this.state;
+  return newState;
+};
 
-    if (OBJECT_TYPES.indexOf(searchTerm) < 0) return;
-    const data = database.objects(searchTerm);
-    this.unfilteredData = data;
-    this.setState({
-      data,
-      dataSource: dataSource.cloneWithRows(data),
-    });
-  }
+/**
+ * A page for displaying objects in the local database. Includes search and filtering functionality.
+ *
+ * @prop   {UIDatabase}     database       App wide database.
+ * @state  {string}         objectString   Current database object.
+ * @state  {string}         searchString   Current search string. Used to update current object.
+ * @state  {string}         filterString   Current filter string. Used to update filtered data.
+ * @state  {Realm.Results}  objectData     Reference to current database object results. Used to
+ *                                         roll back filter state when filter is reset.
+ * @state  {Realm.Results}  filteredData   Reference to current database object results after filter
+ *                                         has been applied. Displayed to the user.
+ */
+export const RealmExplorer = ({ database }) => {
+  const [state, setState] = useState(getInitialState(database));
 
-  renderHeader() {
-    const { data } = this.state;
-
-    const headerCells = [];
-    if (data && data.length > 0) {
-      const firstObject = data[0];
-
-      // TODO: use explicit loops instead of generators.
-      // eslint-disable-next-line no-restricted-syntax
-      for (const [key] of Object.entries(firstObject)) {
-        headerCells.push(
-          <HeaderCell
-            key={key}
-            style={globalStyles.headerCell}
-            textStyle={globalStyles.text}
-            width={1}
-            text={key}
-          />
-        );
-      }
-    }
-    return <Header style={globalStyles.header}>{headerCells}</Header>;
-  }
+  const onSearchChange = searchString => {
```

Comment:
the way this is done there's two renders per character type. (when you use internal `SearchBar` it will auto debounce) - but will still cause two renders per update